# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [98]:
# Imports section

## 1. Loading the dataset

In [99]:
import pandas as pd

# Using pandas load the dataset
data = pd.read_csv("science_data_large.csv")
# Output the first 15 rows of the data
print (data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [100]:
from sklearn.model_selection import train_test_split
# Take the pandas dataset and split it into our features (X) and label (y)
x = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=42)
print(x_train)
print(x_test)
print(y_train)
print(y_test)
# For grading consistency use random_state=42 

     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]
     Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595
..              ...       ...
436             421       725
764             431       860
88              983       433
63              580       817
826              24        89

[100 rows x 2 columns]
716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
           ...     
106     40433.71429
270    335545.00000
860    600674.85710
435    441286.60000
102    505469.25710
Name: Size n

## 3. Perform a Linear Regression

In [101]:
from sklearn.linear_model import LinearRegression
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(x_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
datapoint = [[132, 777]]
predicted = model.predict(datapoint)

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(x_test, y_test)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(coefficients)
print(intercept)
print(predicted)
print(score)

[ 866.14641337 1032.69506649]
-409391.47958340764
[507343.91364115]
0.8552472077276095


/home/codespace/.local/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

h(x) = -409391.47958340764 + 866.14641337(Temperature) + 1032.69506649(KCL)

Report on score and explain meaning: An R^2 score of 0.855 indicates that the model's predictions are decently accurate. An R^2 score of 1 represents a perfect prediction, while a score of 0 suggests the model may not be the best fit for the data trends. Therefore, an R^2 score of 0.855 means the model is providing predictions that are close to the actual values and performing well, but there's still some room for improvement.

## 4. Use Cross Validation

In [102]:
from sklearn.model_selection import cross_val_score, KFold
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model, x, y, cv=kf)

print(cv_scores)
# Report on their finding and their significance

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here: The five scores, all above 0.8, show that the model remains accurate across different data shuffles, indicating a strong performance. The small variance between the scores highlights the model's stability and consistency. Cross-validation also ensures that the model isn't overfitting to a single test set, as it is evaluated on multiple subsets, confirming that it generalizes well to unseen data.

It seems like the model retains accuracy despite changes in the training and testing sets. 

## 5. Using Polynomial Regression

In [103]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score

# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(x)
X_train_poly, X_test_poly, Y_train, Y_test = train_test_split(X_poly, y, test_size=0.1, random_state=42)
reg_poly = LinearRegression()
reg_poly.fit(X_train_poly, Y_train)

# Perform k-fold cross validation (as above)
kfo = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores_poly = cross_val_score(reg_poly, X_poly, y, cv=kfo)
# Report on the metrics and output the resultant equation as you did in Part 3.

datapoint2 = [[100, 100]]
datapoint2_poly = poly.transform(datapoint2)
predicted2 = reg_poly.predict(datapoint2_poly)
score2 = reg_poly.score(X_test_poly, Y_test)
intercept2 = reg_poly.intercept_
coefficients2 = reg_poly.coef_

print(coefficients2)
print(intercept2)
print(score2)
print(predicted2)
print(cv_scores_poly)

[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
2.0477105863392353e-05
1.0
[21485.71429291]
[1. 1. 1. 1. 1.]


/home/codespace/.local/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names
  warnings.warn(


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

h(x) = (2.0477*10^-5) + 12(Temperature) + (−1.27195488 * 10^-7) (KCL) + (1.26494371 * 10^-11)(Temperature^2) + 2(KCL^2) + (2.85714287 * 10^-2)(Temperature)(KCL) 

Report on the score and interpret: A score of 1 indicates a perfect prediction. Since the model consistently achieves this score across multiple shuffles in test sets, it demonstrates that it reliably makes accurate predictions every time.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
# Load the dataset. Then train and evaluate the classification models.

kidney_data = pd.read_csv("ckd_feature_subset.csv")

x = kidney_data[['age', 'bp', 'wbcc', 'appet_poor', 'appet_good', 'rbcc']]
y = kidney_data['Target_ckd']

A_train, A_test, B_train, B_test = train_test_split(x, y, test_size=0.1, random_state=42)

log_reg = LogisticRegression()
log_reg.fit(A_train, B_train)

svm = SVC()
svm.fit(A_train, B_train)

knn = KNeighborsClassifier()
knn.fit(A_train, B_train)

nn = MLPClassifier()
nn.fit(A_train, B_train)

kfol = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores1 = cross_val_score(log_reg, x, y, cv=kfol)
cv_scores2 = cross_val_score(svm, x, y, cv=kfol)
cv_scores3 = cross_val_score(knn, x, y, cv=kfol)
cv_scores4 = cross_val_score(nn, x, y, cv=kfol)

print(cv_scores1)
print(cv_scores2)
print(cv_scores3)
print(cv_scores4)

print("Logistic Regression: Mean =", np.mean(cv_scores1), ", Std Dev =", np.std(cv_scores1))
print("SVM: Mean =", np.mean(cv_scores2), ", Std Dev =", np.std(cv_scores2))
print("KNN: Mean =", np.mean(cv_scores3), ", Std Dev =", np.std(cv_scores3))
print("Neural Network: Mean =", np.mean(cv_scores4), ", Std Dev =", np.std(cv_scores4))



/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
[0.90322581 1.         0.87096774 0.9        0.96666667]
[0.93548387 1.         0.87096774 0.86666667 0.96666667]
[0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
Logistic Regression: Mean = 0.8565591397849461 , Std Dev = 0.0662686697755876
SVM: Mean = 0.9281720430107526 , Std Dev = 0.04760126519964333
KNN: Mean = 0.9279569892473118 , Std Dev = 0.05243961560473918
Neural Network: Mean = 0.8565591397849461 , Std Dev = 0.0662686697755876


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

| Model | Mean Accuracy | Standard Deviation |
|---------------------|---------------|---------------------|
| Logistic Regression | 0.8566 | 0.0663 |
| Support Vector Machine (SVM) | 0.9282 | 0.0476 |
| k-Nearest Neighbors (KNN) | 0.9280 | 0.0524 |
| Neural Network | 0.8566 | 0.0663 |

In this evaluation, the Support Vector Machine (SVM) model demonstrated the highest performance, achieving a mean accuracy of 0.928, which exceeds the accuracy of the other models assessed. This superior performance can be attributed to SVM's ability to effectively handle both linear and non-linear relationships, particularly in high-dimensional datasets. By utilizing kernel functions to maximize the margin between classes, SVM is capable of generalizing well, as evidenced by its low standard deviation of 0.0476, indicating consistent performance across different data splits.
In comparison, the k-Nearest Neighbors (KNN) model also performed admirably, with a mean accuracy of 0.9279 and a slightly higher standard deviation of 0.0524. This suggests that while KNN is quite effective, it may exhibit slightly greater variability in its performance across different subsets of the data.
Other models, such as Logistic Regression and Neural Networks, achieved lower accuracies, both with a mean of 0.8566 and higher standard deviations (0.0663), indicating less stability. The limitations in these models may stem from their inability to capture complex patterns inherent in the data or manage its high dimensionality effectively.

In [106]:
# Experiments with Neural Network.
from sklearn.model_selection import KFold, cross_val_score
import numpy as np

nn1 = MLPClassifier(hidden_layer_sizes=(50,), activation='relu', max_iter=1000)
nn2 = MLPClassifier(hidden_layer_sizes=(50,), activation='logistic', max_iter=10000)
nn3 = MLPClassifier(hidden_layer_sizes=(500,), activation='logistic', max_iter=1000)

kfol = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores_nn1 = cross_val_score(nn1, x, y, cv=kfol)
cv_scores_nn2 = cross_val_score(nn2, x, y, cv=kfol)
cv_scores_nn3 = cross_val_score(nn3, x, y, cv=kfol)

print(cv_scores_nn1)
print(cv_scores_nn2)
print(cv_scores_nn3)

print("NN1: Mean =", np.mean(cv_scores_nn1), ", Std Dev =", np.std(cv_scores_nn1))
print("NN2: Mean =", np.mean(cv_scores_nn2), ", Std Dev =", np.std(cv_scores_nn2))
print("NN3: Mean =", np.mean(cv_scores_nn3), ", Std Dev =", np.std(cv_scores_nn3))



[0.93548387 0.96774194 0.87096774 0.96666667 0.96666667]
[0.87096774 0.96774194 0.83870968 0.9        0.96666667]
[0.90322581 0.96774194 0.83870968 0.76666667 0.96666667]
NN1: Mean = 0.9415053763440862 , Std Dev = 0.03732654415251637
NN2: Mean = 0.9088172043010753 , Std Dev = 0.05146663863023324
NN3: Mean = 0.8886021505376345 , Std Dev = 0.07736736605674542


## Results and Conclusion for Neural Network Experiments

| Neural Network Configuration | Mean Accuracy | Standard Deviation |
|-------------------------------|---------------|---------------------|
| NN1 | 0.9415 | 0.0373 |
| NN2 | 0.9088 | 0.0515 |
| NN3 | 0.8886 | 0.0774 |

The performance evaluation of three neural network configurations reveals that the first configuration (NN1), utilizing a ReLU activation function with 50 neurons, and 1000 max iterations achieved the highest mean accuracy of 0.9415, indicating its effectiveness. The second configuration (NN2), which also employed 50 neurons and a max iteration of 10000 but with a logistic activation function, demonstrated a respectable mean accuracy of 0.9088, though it exhibited a higher standard deviation of 0.0515, suggesting less stability across different folds. Conversely, the third configuration (NN3), utilizing a logistic activation function with 500 neurons and 1000 max iterations resulted in the lowest mean accuracy of 0.8886 and the highest standard deviation of 0.0774, indicating potential overfitting. Overall, the results show that in this instance the specific activation function plays a larger role in the score of a model then the number of neurons or the max iteration. In addtion, the number of max iterations is also has a larger effect on the positive score of the model when compared to the number of Neurons. However, there is addtional nuances as if I added 100000000 max iterations, it could change the result so that the max iteration has the greatest positive effect on score. 